### DISCUSSION TITLE
Unstable NequIP potential using dataset that worked with PaiNN

### DISCUSSION THREAD

**Discussion by gavinwinter** [mir-group/nequip#219] [Q&A] [answered]
I am running NVT with the NequIP potential, and it seems like the disorder in the supercell is causing problems, where specific atoms just get locked in a site and things start freezing (after only 10ps), With the same dataset, training a potential with PaiNN seems to give stable MD for nanoseconds, but with the same dataset training a potential with NequIP seems pretty unstable (at least with off-the-shelf hyperparameters). Any thoughts on what the issue may be? Updating the neighbor list every 5 steps seems pretty reasonable, as with PaiNN we can get away with updating it every 20 steps.

I have attached the LAMMPS input file, but I think the parameters for MD are pretty standard.

```
    # ---------- Initialize simulation ---------------------
    units real
    atom_style atomic
    newton off
    dimension  3
    boundary   p p p
    read_data /home/gridsan/gawinter/projects/labutil/lammps.data
    
    pair_style nequip
    pair_coeff * * /home/gridsan/gawinter/projects/other_nff/results/lgps/example-lgps/lgps-deployed.pth  Li Ge P S
    mass            1 6.941
    mass            2 72.59
    mass            3 30.97376
    mass            4 32.06

    neighbor 1.0 bin
    neigh_modify delay 5 every 1

    velocity  all create 300 87287 dist gaussian
    # ---------- Describe computed properties------------------
    thermo_style custom step pe ke etotal temp
    thermo 100
    dump 1 all custom 20 dump.md.1.* id type xs ys zs
    # ---------- Specify ensemble  ---------------------
    # fix  1 all nve
    fix 1 all nvt temp 300 300 50
    # --------- Run -------------
    timestep 1
    run 100000

```

![nequip_top](https://user-images.githubusercontent.com/42705484/172489865-da50a0a8-af77-4458-b697-46185ef70552.gif)
![nequip_front](https://user-images.githubusercontent.com/42705484/172489867-ef7cfb53-3a3b-4daa-b5a5-be901ec52338.gif)

**Comment by simonbatzner** [comment, marked as answer]
Can you share the NequIP config file. I see units real in LAMMPS, are your data in kcal/mol? The comparison to PaiNN doesn't tell you much, this looks like something is just set up wrong or unconverged, PaiNN is a subset of NequIP and we have consistently seen better results.

**Reply by gavinwinter** [reply]
Agreed. What we get with NequIP should be as good as PaiNN. That's why I suspect I'm doing something wrong. I attached my NequIP config file. Thanks for the help.

```
# general
root: results/lgps
run_name: example-lgps
seed: 123
dataset_seed: 456

# network
r_max: 4.0                                                                        # cutoff radius in length units, here Angstrom, this is an important hyperparamter to scan
num_layers: 4                                                                     # number of interaction blocks, we find 3-5 to work best
l_max: 1                                                                          # the maximum irrep order (rotation order) for the network's features, l=1 is a good default, l=2 is more accurate but slower
parity: true                                                                      # whether to include features with odd mirror parity; often turning parity off gives equally good results but faster networks, so do consider this
num_features: 32                                                                  # the multiplicity of the features, 32 is a good default for accurate network, if you want to be more accurate, go larger, if you want to be faster, go lower
nonlinearity_type: gate                                                           # may be 'gate' or 'norm', 'gate' is recommended

# scalar nonlinearities to use â€” available options are silu, ssp (shifted softplus), tanh, and abs.
# Different nonlinearities are specified for e (even) and o (odd) parity;
# note that only tanh and abs are correct for o (odd parity)
# silu typically works best for even 
nonlinearity_scalars:
  e: silu
  o: tanh

nonlinearity_gates:
  e: silu
  o: tanh

# radial network basis
num_basis: 8                                                                      # number of basis functions used in the radial basis, 8 usually works best
BesselBasis_trainable: true                                                       # set true to train the bessel weights
PolynomialCutoff_p: 6                                                             # p-exponent used in polynomial cutoff function, smaller p corresponds to stronger decay with distance

# radial network
invariant_layers: 2                                                               # number of radial layers, usually 1-3 works best, smaller is faster
invariant_neurons: 64                                                             # number of hidden neurons in radial function, smaller is faster
avg_num_neighbors: auto                                                           # number of neighbors to divide by, null => no normalization, auto computes it based on dataset 
use_sc: true                                                                      # use self-connection or not, usually gives big improvement

# data set
# the keys used need to be stated at least once in key_mapping, npz_fixed_field_keys or npz_keys
# key_mapping is used to map the key in the npz file to the NequIP default values (see data/_key.py)
# all arrays are expected to have the shape of (nframe, natom, ?) except the fixed fields
# note that if your data set uses pbc, you need to also pass an array that maps to the nequip "pbc" key
dataset: npz                                                                       # type of data set, can be npz or ase
dataset_file_name: ./datasets/lgps_dft.npz                                        # path to data set file
key_mapping:
  pbc: pbc
  cell: cell                                                                       # cell
  z: atomic_numbers                                                                # atomic species, integers
  E: total_energy                                                                  # total potential eneriges to train to
  F: forces                                                                        # atomic forces to train to
  R: pos                                                                           # raw atomic positions
npz_fixed_field_keys:                                                              # fields that are repeated across different examples
  - atomic_numbers

chemical_symbols:
  - Li
  - Ge
  - P
  - S

# logging
wandb: false                                                                        # we recommend using wandb for logging


verbose: info                                                                      # the same as python logging, e.g. warning, info, debug, error; case insensitive
log_batch_freq: 10                                                                 # batch frequency, how often to print training errors withinin the same epoch
log_epoch_freq: 1                                                                  # epoch frequency, how often to print 
save_checkpoint_freq: -1                                                           # frequency to save the intermediate checkpoint. no saving of intermediate checkpoints when the value is not positive.
save_ema_checkpoint_freq: -1                                                       # frequency to save the intermediate ema checkpoint. no saving of intermediate checkpoints when the value is not positive.

# training
n_train: 100                                                                       # number of training data
n_val: 50                                                                          # number of validation data
learning_rate: 0.005                                                               # learning rate, we found values between 0.01 and 0.005 to work best - this is often one of the most important hyperparameters to tune
batch_size: 5                                                                      # batch size, we found it important to keep this small for most applications including forces (1-5); for energy-only training, higher batch sizes work better
max_epochs: 100                                                                    # stop training after _ number of epochs, we set a small number here to have an example that finished within a few minutes, but in practice we recommend using a very large number, as e.g. 1million and then to just use early stopping and not train the full number of epochs
train_val_split: random                                                            # can be random or sequential. if sequential, first n_train elements are training, next n_val are val, else random, usually random is the right choice
shuffle: true                                                                      # if true, the data loader will shuffle the data, usually a good idea
metrics_key: validation_loss                                                       # metrics used for scheduling and saving best model. Options: `set`_`quantity`, set can be either "train" or "validation, "quantity" can be loss or anything that appears in the validation batch step header, such as f_mae, f_rmse, e_mae, e_rmse
use_ema: true                                                                      # if true, use exponential moving average on weights for val/test, usually helps a lot with training, in particular for energy errors
ema_decay: 0.99                                                                    # ema weight, typically set to 0.99 or 0.999
ema_use_num_updates: true                                                          # whether to use number of updates when computing averages
report_init_validation: true                                                       # if True, report the validation error for just initialized model

# early stopping based on metrics values.
early_stopping_patiences:                                                          # stop early if a metric value stopped decreasing for n epochs
  validation_loss: 50

early_stopping_lower_bounds:                                                       # stop early if a metric value is lower than the bound
  LR: 1.0e-5

# loss function
loss_coeffs:                                                                        
  forces: 1                                                                        # if using PerAtomMSELoss, a default weight of 1:1 on each should work well
  total_energy:                                                                    
    - 1
    - PerAtomMSELoss

# output metrics
metrics_components:
  - - forces                               # key 
    - mae                                  # "rmse" or "mae"
  - - forces
    - rmse
  - - forces
    - mae
    - PerSpecies: True                     # if true, per species contribution is counted separately
      report_per_component: False          # if true, statistics on each component (i.e. fx, fy, fz) will be counted separately
  - - forces                                
    - rmse                                  
    - PerSpecies: True                     
      report_per_component: False    
  - - total_energy
    - mae    
  - - total_energy
    - mae
    - PerAtom: True                        # if true, energy is normalized by the number of atoms

# optimizer, may be any optimizer defined in torch.optim
# the name `optimizer_name`is case sensitive
optimizer_name: Adam                                                               # default optimizer is Adam 
optimizer_amsgrad: false

# lr scheduler, currently only supports the two options listed in full.yaml, i.e. on-pleteau and cosine annealing with warm restarts, if you need more please file an issue
# here: on-plateau, reduce lr by factory of lr_scheduler_factor if metrics_key hasn't improved for lr_scheduler_patience epoch
lr_scheduler_name: ReduceLROnPlateau
lr_scheduler_patience: 100
lr_scheduler_factor: 0.5

# we provide a series of options to shift and scale the data
# these are for advanced use and usually the defaults work very well
# the default is to scale the atomic energy and forces by scaling them by the force standard deviation and to shift the energy by the mean atomic energy
# in certain cases, it can be useful to have a trainable shift/scale and to also have species-dependent shifts/scales for each atom

# whether the shifts and scales are trainable. Defaults to False. Optional
per_species_rescale_shifts_trainable: false
per_species_rescale_scales_trainable: false

# initial atomic energy shift for each species. default to the mean of per atom energy. Optional
# the value can be a constant float value, an array for each species, or a string that defines a statistics over the training dataset
per_species_rescale_shifts: dataset_per_atom_total_energy_mean

# initial atomic energy scale for each species. Optional.
# the value can be a constant float value, an array for each species, or a string
per_species_rescale_scales: dataset_forces_rms

# if explicit numbers are given for the shifts/scales, this parameter must specify whether the given numbers are unitless shifts/scales or are in the units of the dataset. If ``True``, any global rescalings will correctly be applied to the per-species values.
# per_species_rescale_arguments_in_dataset_units: True
```

**Reply by gavinwinter** [reply]
Realized I never answered your other question. Yes, my dataset units are in kcal/mol and kcal/mol-Angs (to agree with units real).

**Reply by simonbatzner** [reply]
You're only training for 100 epochs, so your training is probably not converged. Note that comment in `configs/example.yaml`: 


```
max_epochs: 100                                                                    # stop training after _ number of epochs, we set a small number here to have an example that finished within a few minutes, but in practice we recommend using a very large number, as e.g. 1million and then to just use early stopping and not train the full number of epochs`
```

Set this to a very high number and then stop your training when the LR has dropped past a certain value, e.g. 1e-5, or the validation loss has not improved for _ epochs, e.g. 50 or 100: 

```
# early stopping based on metrics values.
early_stopping_patiences:                                                          # stop early if a metric value stopped decreasing for n epochs
  validation_loss: 50

early_stopping_lower_bounds:                                                       # stop early if a metric value is lower than the bound
  LR: 1.0e-5
```

**Reply by gavinwinter** [reply]
Thanks, Simon. I’ll try that and do a convergence check along with a test evaluation. I’ve just previously seen convergence/plateau within 100 epochs for training PaiNN and SchNet, so I figured it was fine. I’ll check more carefully for convergence this time.

**Reply by gavinwinter** [reply]
That was part of the problem, yes. In addition, I did not read things carefully enough: I interpreted `n_train` to be something like the batch size, so I was only training on 100 geometries out of my dataset of ~28,000. (1) I set `n_train` to 17,000 (so as to have a reasonable train-val-test split) and (2) set the `max_epochs` to a high number, and everything seems good now. Training did not plateau by the convergence criterion I set, but I stopped training after 1000 epochs, and the potential seems stable for at least 100ps now and is probably good for more. So, the issue seems to be resolved. Thanks.

**Reply by simonbatzner** [reply]
I see, yeah that'll do it. Just make sure to not use a batch size of 100, we found smaller ones work much better. And good to hear things are looking good now.
